### DISCUSSION TITLE
Protected branch and deploy keys

### DISCUSSION THREAD

**Discussion by jtognazzi** [python-gitlab/python-gitlab#2366] [Q&A] [answered]
Hi everybody,

python-gitlab does not seems to support allowing a deploy key to push to a protected branch.
Is this planned ?
Looking at the http requests when performing the task via the WebUI,
it seems it is done using `push_access_levels` parameter, like this:
```
...
	"push_access_levels": [
		{
			"id": 1665,
			"protected_branch_id": 1635,
			"access_level": 0,
			"created_at": "2022-11-07T16:15:38.432+01:00",
			"updated_at": "2022-11-07T16:15:38.432+01:00",
			"user_id": null,
			"group_id": null,
			"deploy_key_id": null
		},
		{
			"id": 1667,
			"protected_branch_id": 1635,
			"access_level": 40,
			"created_at": "2022-11-07T16:34:41.214+01:00",
			"updated_at": "2022-11-07T16:34:41.214+01:00",
			"user_id": null,
			"group_id": null,
			"deploy_key_id": 333
		}
	]
```

I wanted to experiment what is possible with push_access_levels and the CLI,
but I can't find out how to use this parameter...
```
gitlab --gitlab my-api-key project-protected-branch create --project-id 'myproject' --name 'toto/*' --push-access-level "[{'access_level': 40, 'user_id': 248}]"
Impossible to create object (400: push_access_level is invalid, push_access_level does not have a valid value)
```

I found some issue about this, but I'm still quite lost here:
https://github.com/python-gitlab/python-gitlab/issues/1732
https://github.com/python-gitlab/python-gitlab/pull/1699

Thanks for any hint

**Comment by nejch** [comment, marked as answer]
Hi @jtognazzi this is currently not supported by the CLI yet as we need to first support parsing arrays of hashes, as mentioned in the other issue and elsewhere. You can do this via the API though.
